File: fieldmark_py/fieldmark/authz.py
"""Authorization decision primitive (FR5).

`can(user, action, entity_id=None) -> bool` is the single Django-side call
site every view uses to decide whether an action is permitted. Epic 1:
role-only checks (entity-scope rules deferred to Epic 2+).
"""

from __future__ import annotations

import uuid
from typing import Any

from fieldmark.roles import Role

# Action → frozenset of permitted role names. Epic 1 ships empty;
# subsequent stories register their actions via register_action().
_ACTION_ROLE_MAP: dict[str, frozenset[str]] = {}


def register_action(action: str, *roles: Role) -> None:
    """Register an action → permitted-roles mapping. Call at app-load
    time (typically a module-level statement in the same package as the
    action's handler — Django signals are banned, so AppConfig.ready()
    is off-limits).
    """
    existing = _ACTION_ROLE_MAP.get(action, frozenset())
    _ACTION_ROLE_MAP[action] = existing | frozenset(r.value for r in roles)


def can(user: Any, action: str, entity_id: uuid.UUID | None = None) -> bool:
    if not getattr(user, "is_authenticated", False):
        return False
    permitted = _ACTION_ROLE_MAP.get(action)
    if permitted is None:
        return False
    user_roles = set(user.groups.values_list("name", flat=True))
    if not (user_roles & permitted):
        return False
    return _evaluate_entity_scope(action, entity_id)


def _evaluate_entity_scope(action: str, entity_id: uuid.UUID | None) -> bool:
    # Single extension point for Epic 2+ entity-scope rules. Today every
    # action is role-coarse; future stories swap this for per-action
    # entity-scope evaluators.
    return True


# Test-only escape hatch. Production callers must use register_action().
def _reset_for_tests() -> None:
    _ACTION_ROLE_MAP.clear()

File: fieldmark-go/internal/web/auth/authz.go
package auth

import (
	"sync"

	"github.com/gofiber/fiber/v3"
	"github.com/google/uuid"

	"github.com/code-chimp/fieldmark-go/internal/app"
	"github.com/code-chimp/fieldmark-go/internal/domain"
)

// actionRoleMap holds the action → permitted-roles mapping. Stories from
// Epic 2+ register their actions at composition time. Story 1.12 ships
// empty — Epic 1 has no live actions.
var (
	actionRoleMapMu sync.RWMutex
	actionRoleMap   = map[string]map[domain.Role]struct{}{}
)

// RegisterAction registers an action → permitted-roles mapping.
func RegisterAction(action string, roles ...domain.Role) {
	actionRoleMapMu.Lock()
	defer actionRoleMapMu.Unlock()
	set, ok := actionRoleMap[action]
	if !ok {
		set = map[domain.Role]struct{}{}
		actionRoleMap[action] = set
	}
	for _, r := range roles {
		set[r] = struct{}{}
	}
}

// Can returns true if the actor is authenticated and permitted to perform
// action (optionally scoped to entityID; uuid.Nil means "no entity scope").
// Epic 1: role-only checks (entity-scope rules deferred to Epic 2+).
func Can(actor *app.Actor, action string, entityID uuid.UUID) bool {
	if actor == nil || actor.IsAnonymous() {
		return false
	}
	actionRoleMapMu.RLock()
	permitted, ok := actionRoleMap[action]
	actionRoleMapMu.RUnlock()
	if !ok {
		return false
	}
	if _, hit := permitted[domain.Role(actor.Role)]; !hit {
		return false
	}
	return evaluateEntityScope(action, entityID)
}

// evaluateEntityScope is the single extension point for Epic 2+ entity-scope
// rules (e.g., "Site Supervisor can act on a Violation only if assigned to it").
// Today every action is role-coarse; future stories wire entity-scope evaluators here.
func evaluateEntityScope(_ string, _ uuid.UUID) bool { return true }

// resetForTests clears the map. Test-only — unexported, accessible only within this package.
func resetForTests() {
	actionRoleMapMu.Lock()
	defer actionRoleMapMu.Unlock()
	actionRoleMap = map[string]map[domain.Role]struct{}{}
}

// ResetForTests clears the action policy map for external package tests.
func ResetForTests() {
	resetForTests()
}

// RequireRole returns a middleware that 403s if the authenticated actor does
// not hold the given conceptual role. Unauthenticated actors are redirected to
// /login. Use the domain.Role* constants for the role argument.
func RequireRole(role string) fiber.Handler {
	return func(c fiber.Ctx) error {
		actor := ActorFromCtx(c)
		if actor.IsAnonymous() {
			return c.Redirect().Status(fiber.StatusFound).To("/login")
		}
		if actor.Role != role {
			c.Status(fiber.StatusForbidden)
			return c.SendString("Forbidden.")
		}
		return c.Next()
	}
}
